neo/Consensus/ConsensusService.cs

```diff
@@ -228,6 +256,201 @@ private void OnPersistCompleted(Block block)
             InitializeConsensus(0);
         }
 
+        private bool ReverifyPrepareRequest(ConsensusContext consensusContext, RecoveryMessage message, Snapshot snapshot,
+            out ConsensusPayload prepareRequestPayload, out PrepareRequest prepareRequest)
+        {
+            if (message.TransactionHashes != null)
+            {
+                prepareRequest = new PrepareRequest
+                {
+                    ViewNumber = consensusContext.ViewNumber,
+                    Nonce = message.Nonce,
+                    NextConsensus = message.NextConsensus,
+                    TransactionHashes = message.TransactionHashes,
+                    MinerTransaction = message.MinerTransaction
+                };
+                var prepareRequestTimestamp = message.PrepareTimestamps[consensusContext.PrimaryIndex];
+                prepareRequestPayload = consensusContext.RegenerateSignedPayload(
+                    prepareRequest, (ushort) consensusContext.PrimaryIndex,
+                    message.PrepareWitnessInvocationScripts[consensusContext.PrimaryIndex], prepareRequestTimestamp);
+
+                if (prepareRequestPayload.Verify(snapshot) && PerformBasicConsensusPayloadPreChecks(prepareRequestPayload))
+                    return true;
+            }
+
+            prepareRequestPayload = null;
+            prepareRequest = null;
+            return false;
+        }
+
+        private void HandleRecoveryInCurrentView(RecoveryMessage message, Snapshot snap)
+        {
+            if (context.State.HasFlag(ConsensusState.BlockSent)) return;
+
+            ConsensusPayload prepareRequestPayload = null;
+            UInt256 preparationHash;
+            // if we are already on the right view number the only thing we might want to do is accept more
+            // Preparation messages if they can be reconstructed from the regeneration message.
+            if (context.PrimaryIndex == context.MyIndex)
```

Comment:
We need to double check this case as well.

There was an edge case for when the Primary was being regenerated.
* the Primary n_1 already picked the Transactions Hashes and then failled;
* After that some of the others nodes entered commit phase;
* Primary came back and should be regenerated. But he need to be regenerated with the same transactions hashes that he picked before. He can not create the PrepareRequest himself.

I am not sure if this is being accomplished here, we were doing that the the other PR by calling  `OnPrepareRequestReceived`  in every case.

Reply:
We need also to be careful with https://github.com/neo-project/neo/blob/e6cc6e4ae1445d2f2e1b5826620dc429fe4327ba/neo/Consensus/ConsensusService.cs#L436

This might be a complex case if other nodes already got PrepareReq from that Primary with a different set of transactions.

Reply:
Yes, I will change this in some coming changes to this PR, so that it will instead accept its previously sent prepare request.

Reply:
> We need also to be careful with https://github.com/neo-project/neo/blob/e6cc6e4ae1445d2f2e1b5826620dc429fe4327ba/neo/Consensus/ConsensusService.cs#L436

This code has been changed now and it re-accepts it's previously generated prepare request.

Reply:
I think my last TODO I just added is still related to resolving this conversation. Going to work on it now, and then it needs to be tested and again reviewed closely.

Reply:
This has been resolved. The primary will now restore it's own previously generated view and act as a backup. Feel free to unresolve,  if you believe something is not resolved here.